users/test.py

```diff
@@ -36,36 +40,136 @@ def setUp(self):
 
 
+class TestUserUpsertion(unittest.TestCase):
+    """Test updating/inserting users into the db."""
+
+    def setUp(self):
+        """Seed mock DB for testing"""
+        self.mock_collection = mongomock.MongoClient().db.collection
+
+    def test_insert_valid_user(self):
+        """A valid user object should be upserted and retrieved correctly.
+
+        Checks the user found after upsertion both against the original
+        object added to the database and against the object found with
+        the ObjectID returned from app.upsert_user_in_db.
+        """
+        user_to_insert = {
+            "user_id": USER_ID,
+            "name": USER_NAME
+        }
+        upserted_id = app.upsert_user_in_db(
+            user_to_insert, self.mock_collection)
+        found_user = self.mock_collection.find_one({"user_id": USER_ID})
+        # original user attributes matches found user attributes
+        self.assertEqual(user_to_insert["user_id"], found_user["user_id"])
+        self.assertEqual(user_to_insert["name"], found_user["name"])
+        # user returned from app.upsert_user_in_db matches found user exactly
+        returned_user = self.mock_collection.find_one(upserted_id)
+        self.assertEqual(returned_user, found_user)
+
+    def test_user_auth_defaults_are_set(self):
+        """app.upsert_user_in_db should set any default value before upsertion.
+
+        Right now, only the 'is_organizer' field of an inserted user should
+        default to False.
+        """
+        user_to_insert = {
+            "user_id": USER_ID,
+            "name": USER_NAME
+        }
+        app.upsert_user_in_db(user_to_insert, self.mock_collection)
+        found_user = self.mock_collection.find_one({"user_id": USER_ID})
+        self.assertFalse(found_user["is_organizer"])
+
+    def test_malformatted_user_not_inserted(self):
+        """A malformatted user_object should not be inserted.
+
+        Note: app.upsert_user_in_db should not insert the user if it is missing
+        the 'user_id' or 'name' attributes, but should insert the user if
+        the user has more attributes than those. If the user is malformatted,
+        app.upsert_user_in_db raises an AttributeError, else it upserts the
+        user and returns the new user object.
+        """
+        # missing name
+        with self.assertRaises(AttributeError):
+            app.upsert_user_in_db({"user_id": USER_ID}, self.mock_collection)
+        # missing user_id
+        with self.assertRaises(AttributeError):
+            app.upsert_user_in_db({"name": USER_NAME}, self.mock_collection)
+        # missing both
+        with self.assertRaises(AttributeError):
+            app.upsert_user_in_db({}, self.mock_collection)
+        # no users should have been inserted
+        self.assertEqual(self.mock_collection.count_documents({}), 0)
+
+    def test_additional_info_still_inserted(self):
+        """A user_object with additional information should still be inserted.
```

Comment:
Do you ever plan to use any of the extra info? If not, might be best to not insert it in order to prevent too much extra data being stored.

Reply:
I was thinking it would be useful to easily add email or phone info later for notifications which might be in some but not all users, but I guess we don't need that now so maybe it's best left out. @evankanderson ?

Reply:
You should agree whether you want to be strict or lenient here. 

It sounds like we should go the strict route to start with.

Reply:
Done https://github.com/knative-portability/large-events/pull/8/commits/80bbf7263589bdc41a637f599b31f281c314b6f5